src/reference-manual/pedantic-mode.Rmd

```diff
@@ -0,0 +1,345 @@
+      Argument 100 suggests there may be parameters that are not unit scale;
+      consider rescaling with a multiplier (see manual section 22.12).
+
+
+## Control flow depends on a parameter
+
+Control flow statements, such as `if`, `for` and `while`, should not depend on the value of the parameters to determine their branching conditions.
+This is likely to introduce discontinuity into the density function.
+Pedantic mode generates a warning when any branching condition may depend on a parameter value.
+
+For example, in the following program:
+
+    parameters {
+      real a;
+    }
+    model {
+      // x depends on parameter a
+      real x = a * a;
+    
+      int m;
+    
+      // the if-then-else depends on x which depends on a
+      if(x > 0) {
+        //now m depends on x which depends on a
+        m = 1;
+      } else {
+        m = 2;
+      }
+    
+      // for loop depends on m -> x -> a
+      for (i in 0:m) {
+        a ~ normal(i, 1);
+      }
+    }
+
+This produces the following warning:
+
+    Warning at 'param-dep-cf-warn.stan', line 11, column 2 to line 16, column 3:
+      A control flow statement depends on parameter(s): a.
+    Warning at 'param-dep-cf-warn.stan', line 19, column 2 to line 21, column 3:
+      A control flow statement depends on parameter(s): a.
+
+
+## Parameters with multiple twiddles
+
+A warning is generated when a parameter is found on the left-hand side of more than one ~ statements (or an equivalent `target +=` conditional density statement).
+This pattern is not inherently an issue, but it is unusual and may indicate a mistake.
+
+Pedantic mode only searches for repeated statements, it will not for example generate a warning when a ~ statement is executed repeatedly inside of a loop.
+
+For example, in the following program:
+
+    data {
+      real x;
+    }
+    parameters {
+      real a;
+      real b;
+    }
+    model {
+      a ~ normal(0, 1);
+      a ~ normal(x, 1);
+    
+      b ~ normal(1, 1);
+    }
+
+This produces the following warning:
+
+    Warning at 'multi-twiddle.stan', line 9, column 2 to column 19:
+      The parameter a is on the left-hand side of more than one twiddle
+      statement.
+
+
+## Parameters with zero or multiple priors
+
+A warning is generated when a parameter appears to have greater than or less than one prior distribution factor.
+
+This analysis depends on a [*factor graph*](https://en.wikipedia.org/wiki/Factor_graph) representation of a Stan program. A factor F that depends on a parameter P is called a *prior factor for P* if there is no path in the factor graph from F to any data variable except through P.
+
+For example, in the following program:
+
+    data {
+      real x;
+    }
+    parameters {
+      real a;
+      real b;
+      real c;
+      real d;
+    }
+    model
+    {
+      a ~ normal(0, 1); // this is a prior
+      x ~ normal(a, 1); // this is not a prior, since data is involved
```

Comment:
Oooh, I see kinda what is going on.

This check makes me a bit nervous, but this isn't something for this release. Parameterizing priors in terms of data hyperpriors is super convenient.

Horseshoe priors do this, for instance.

Look on page 29 of [this](https://arxiv.org/pdf/1707.01694.pdf)

Find this code:
```
real < lower =0 > scale_icept ; # prior std for the intercept
real < lower =0 > scale_global ; # scale for the half -t prior for tau
real < lower =1 > nu_global ; # degrees of freedom for the half -t priors for tau
real < lower =1 > nu_local ; # degrees of freedom for the half - t priors for lambdas
real < lower =0 > slab_scale ; # slab scale for the regularized horseshoe
real < lower =0 > slab_df ; # slab degrees of freedom for the regularized horseshoe
```

All those parameters are data-defined knobs for priors.

Reply:
Right. I was talking with Bob about this as well. The issue is that we don't distinguish syntactically between modeled and modeled "data". Hyperparameters are an issue, another issue is size variables like `int N`. I manually ignore single `int`s for that reason. It would be cool if we could distinguish those somehow.

In the mean time, it's a good idea to add some explanation here.